### ISSUE TITLE
WW2 Pacific 2e Edition - Factory Upgrade can't be placed anywhere

### ISSUE THREAD

**Issue by asvitkine** [triplea-game/triplea#3506]
WW2 Pacific 2e Edition - Factory Upgrade can't be placed anywhere.

At least as Japanese player with having already purchased minor factories in various territories. Either it doesn't work at all - or only some players can place it - in which case Japan shouldn't be allowed to purchase it.

### Engine version
Trunk at git hash ac1015b

### My Operating System
Mac

### Map name
World War II Pacific 1940 2nd Edition

### Can you describe how to trigger the error? (eg: what sequence of actions will recreate it?)
Purchase "Factory Upgrade" as Japan. Try to place it.

### Do you have the exact error text? Please copy/paste if so
N/A

### Instead of this error, what should have happened?
Either the factory upgrade could be placed (e.g. on top of a factory that you've previously purchased) - or it should not be allowed to purchase it.

### Any additional information that may help

Example saved game where I tried to buy it at some point and it couldn't be placed so it just kept being carried-over each round.

[ww2_pac_end.tsvg.zip](https://github.com/triplea-game/triplea/files/2156866/ww2_pac_end.tsvg.zip)

**Comment by ron-murhammer** [comment]
@asvitkine I'm pretty sure the gap here is lack of validation around what you should be allowed to purchase based on if it can be placed. Placement logic is pretty complex across the various maps and rule sets so it would be a decent challenge to improve it to cover all or a larger portion of purchase validation.

**Comment by panther2** [comment]
@asvitkine @ron-murhammer 
Indeed it's a rules issue here: Major IC (or Upgrades) cannot be placed on territories originally controlled by China.
Rulebook:

> The industrial complex to be upgraded must be located on an originally controlled (not captured) territory that you have controlled since the beginning of your turn and that has an IPC value of 3 or higher.

**Comment by panther2** [comment]
Also I would like to add, that from a rules perspective there is no such thing as "what you should be allowed to purchase based on if it can be placed".

The rules allow to purchase "You may buy as many units of any type as you can afford. " during 
Phase 1: Purchase & Repair units

What can be placed is a matter of Phase 5: Mobilize New Units **only**.

I am not voting against any sort of validation. I am just pointing out that it is a "service feature" rather than a rules issue.

**Comment by simon33-2** [comment]
Also a major complex can be placed in Korea. So the issue isn't really specific to Japan. Surprisingly, Triple-A actually blocks you from placing a major on Kiangsu (which is correct). The issue is more that you are allowed to buy an upgrade when you have no upgradeable minors.

Expanding @panther2's point, you aren't supposed to deliberately buy something that you can't place - the rule allowing a refund (not implented in Triple-A) is only supposed to be for inadvertent over purchasing. So it would still be rule compliant if Triple-A blocked purchasing something which couldn't be placed, and in fact more rule compliant than at present and also more compliant than having a warning. I suppose on that last point, I would differ with @panther2.

**Comment by panther2** [comment]
Thank you, @simon33-2 , now I remember we already had that discussion. This is what I answered to this argument in https://github.com/triplea-game/triplea/issues/2219#issuecomment-323506582

> The addition of the prohibition of the "intentional overbuy" into the FAQ did not remove the refund for units that were overbuyed. That refund takes place during the placement phase.
>
>So in case the player purchases (intentionally) more than he can place - and no (other) player realizes this when it occurs (during Purchase Phase), it would be hard to argue for the opponents that this overbuy had been intentional - and to agree about consequences - during Placement Phase.
>
>In other words it would be "compliant behaviour" to not overbuy - and intentionally overbuying hoping that no one realizes this would be sort of cheating.

And here https://github.com/triplea-maps/world_war_ii_global/issues/4#issuecomment-244552740

>Side note: The part of the FAQ dealing with intentional overbuy (=Krieghund's clarification) has not made it into the latest version of the rulebooks (while many of the other parts are included). Not sure what that means... but if this might become important, I will ask him about that.

So indeed, this issue is more or less a duplicate of #2219

**Comment by simon33-2** [comment]
I don't think a warning makes sense from a rules compliance viewpoint. If you override the warning, then haven't you just done an intentional overbuy?

**Comment by panther2** [comment]
With regards of intentionally buying more units than you can place - yes. That is what I said in https://github.com/triplea-maps/world_war_ii_global/issues/4#issuecomment-244552740 
Here the warning is pointless.

With regards to the discussed issue a blocking or warning is not necessary (by the rules) but if someone wants something like this as a feature I would not object, as it does not hurt.

**Comment by Cernelius** [comment]
@panther2 But if in the future TripleA would fully handle purchase validation, giving an always correct warning if you are purchasing too much, wouldn't that realistically count as assuring that any kind of overbuy can only be intentional, thus always forbidden, thus it would be factually the same as if TripleA would never give any warning, nor checking what you are purchasing.

I believe that the only way to be really compliant with the game as intended is to totally remove the warning, and let the players check purchases by themselves, which can already be done, and it might actually be the best choice anyways, in the moment the validation is substantially bugged, map wise.

Personally, I prefer that either something works 100% or not having it. This is why I removed validation completely from 270BC and @veqryn accepted that proposed change/suggestion, to make an example of another map that cannot have a 100% correct validation.

**Comment by panther2** [comment]
@Cernelius I totally agree with your first clause. It would be assuring the intentional overbuy, but as long as it is not pointed out to the opponents, it is pointless. The buying player who intentionally buys too much, can simply ignore a warning and proceed (as long as TripleA does not block).

Meanwhile I have Krieghund's confirmation, that the FAQ-added rule 

`Purchasing Units
Q. Can I purposely purchase more units than I can mobilize with the intent of deciding which
units I want to keep based on the outcome of battles I intend to fight?
A. No. You may only purchase as many units as you will have the ability to mobilize after making repairs
to any damaged industrial complexes. The rules for returning excess units are intended only for
inadvertent over-purchasing.`

included in every FAQ starting from Anniversary Edition up to Europe/PAcific/Global 1940.2 is valid. It was not included into the updated rulebooks but the provided FAQ sheets still apply, including this rule.

The problem now arises due to the wording "inadvertent over-purchasing". In a face-to-face game - ideally - every player would now count what purchases are possible and would discuss this with the buying player. This would most likely prevent from doing intentional over-buys, then. In case no one keeps attention at that point, an intentional overbuy will not become obvious to all before the "Place Units Phase". Then it would be hard to discuss any intention. In the end the buying player would most likely get a refund for an unintentional overbuy.  And a refund would of course be absolutely fine if the player unintentionally bought "too much" and no other player cared or paid attention. That is still the rules, too.

Now in TripleA it would be up to the software "to keep attention" during the Purchase Units Phase and either inform the other players or block it completely, leaving of course no room for (good or bad) "intention".

If TripleA would block overpurchase, the initial issue (buying an IC-upgrade that cannot be placed) discussed here is not related. As this is not an overpurchase but an erroneous purchase of an upgrade that cannot be placed due to territorial rules restrictions. This would however meet the original rulebook-rule for granting a refund, then.

Additionally facilities do not count against the production limits of ICs. So in theory there is no (IC-related) limit for buying facilities (except money and territories to place them).

**Comment by simon33-2** [comment]
I cannot agree with your viewpoint here. Once you override a warning, the overbuy becomes intentional. Unless you believed it to be an incorrect warning, I guess. If you didn't intend to overbuy you can say no.

I wouldn't agree with changing the warning in Global though, because there are a few cases when the warning is indeed incorrect. In spite of this, it adds enough value in the 90+% of cases where it does behave properly to leave the warning there.

All this is getting off the topic though. There are so many cases which would theoretically need to be trapped here once we say that you shouldn't be able to buy something you can't place, which is clearly valid. Just doesn't happen often enough to be that much of a problem. The annoying thing is getting the warning every turn because you can never remove it from the list of unplaced units.

**Comment by panther2** [comment]
> I cannot agree with your viewpoint here. Once you override a warning, the overbuy becomes intentional. 

I am talking about a player **intentionally** overbuying . He can simply ignore the warning, as no one else receives that warning and thus would not notice it. This sort of player would not be stopped by a warning.

All these discussed implications are hard to code in a software as the "human factor" kicks in.

**Comment by panther2** [comment]
> All this is getting off the topic though.

Yes, as I already said here https://github.com/triplea-game/triplea/issues/3506#issuecomment-405297486 , the initial issue is unrelated to the "Overpurchase" discussion, we additionally have here 
https://github.com/triplea-game/triplea/issues/2219

**Comment by simon33-2** [comment]
I don't understand how you consider that an overbuy isn't intentional when the warning is overridden. Anyway, let's agree to disagree.

**Comment by panther2** [comment]
> I don't understand how you consider that an overbuy isn't intentional when the warning is overridden. 

I never said so. 

What I said is that the player who starts his Purchase Phase with the clear intention to overbuy (thus expecting some benefits), will not be stopped by a warning but continue his plan. As his opponents don't get a warning, he can expect or at least hope that they will not notice his intention.


> Anyway, let's agree to disagree.

Haha, agreed :-)

**Comment by simon33-2** [comment]
Oh - perhaps we do agree. Someone doing an intentional overbuy may be able to ignore the warning but he cannot claim that it was inadvertent while the warning is present.

**Comment by panther2** [comment]
Yes. I think overriding the warning serves as (more or less strong) indicator for an intention but not as a final proof **when discussing it after the turn.**

I can think of scenarios where even the good spirited player would override the warning, acting under erroneous assumptions. We still have "The rules for returning excess units are intended only for
inadvertent over-purchasing." in the FAQ.

**Comment by Cernelius** [comment]
>If TripleA would block overpurchase, the initial issue (buying an IC-upgrade that cannot be placed) discussed here is not related. As this is not an overpurchase but an erroneous purchase of an upgrade that cannot be placed due to territorial rules restrictions. This would however meet the original rulebook-rule for granting a refund, then.

Are you sure about this? If I have only 1 upgradable factory and buy 2 factory upgrades, I assumed this would count as an overpurchase too. Is this correct?

If the above is correct, then I don't think that it would make sense to consider differently the case in which I have 0 upgradable factories and I buy any number of factory upgrades.

Also, relatively to the normal placement, thus not necessarily referring to a Global ruleset, in a TripleA perspective, I think we should consider the matter in term of the possibility that there is not a factory in the capital (in a custom game, you design, you don't have to have factories in all your capitals), so you can, as well, having normal units purchases when you have 0 normal placement capabilities. That is currently handled by the program just the same way as any overpurchases.

**Comment by Cernelius** [comment]
>Oh - perhaps we do agree. Someone doing an intentional overbuy may be able to ignore the warning but he cannot claim that it was inadvertent while the warning is present.

>Yes. I think overriding the warning serves as (more or less strong) indicator for an intention but not as a final proof when discussing it after the turn.

I believe you two still disagree...

@panther2 Affirms that you can overbuy and ignore the warning, then argue that you skipped the warning without noticing it, and then you can have the refund; basically saying that in not live TripleA you can always overbuy and get the refund, every single of your turns, as there is no way you can be stopped (nobody yelling at you that you are overpurchasing), while in live TripleA I'm not sure what is his point of view, as the opponent would get to see the purchase choices only after the purchase phase, and, if the turn player is fast enough, that can be little time before getting into combat and related dice rolling (also you can have mods with purchase after move).

@simon33-2 Affirms that skipping the purchase must always count as intentional overbuy, thus making you lose the game, in the moment you cannot place what you bought (skipping the warning would be basically the same as ignoring your opponent yelling at you that you are overpurchasing, during the purchase phase).

Currently, the notes of the game are surely saying that @simon33-2 interpretation is wrong, as you can read this:

>You must place all units you purchased if you are able to, or as many as you can. Any units not placed get refunded to you (TripleA instead carries them over to the next turn).

>Over-purchase is returned and money is reimbursed.

Since you are forbidden to make intentional overpurchases (no matter if you would be fine at not being refunded), if skipping the warning would count as so, then you would never get any refund, thus there would be no reason to have these rules documented in the notes at all, as the refund would in no cases ever actually happen. The fact that they are present is incompatible with @simon33-2 interpretation.

I suggest @ron-murhammer (or any TripleA owners) to make a ruling on this matter (skipping warning counts as intentional overbuy, thus losing the game right away Y/N), documenting it somewhere (no idea where). Then, it may be needed adding it to the notes of each game supposed to work by these rules (intentional overbuy forbidden, but refund granted otherwise), or anyways update such notes if needed.

Of course, the matter would also be solved if TripleA would make impossible to overpurchase, but this would drive it even further away from the intended ruleset, also practically removing the refund rules (not implemented, but at least you can edit), as if TripleA is supposed to support such ruleset, then it should not stop you from making purchase mistakes. On the other hand, since TripleA, for example, is supposed to stop you from making moves mistakes, taking away validation from the opponent eye to the engine, then it would be coherent for it to stop you from making illegal purchases too, which already partially does.

Also @panther2 it has been some years, and haven't checked right now, but I remember that v3 worked the same way (overpurchase allowed only if unintentional), except that you would keep the units, to place them in the future, if possible, instead of getting a refund. If you can confirm this, this basically means that v3 has the same issue, substantially.
Also, I believe that in the previous games, where you would just lose what you can't place, it was not specifically forbidden to make intentional overpurchases, even tho the rules somewhat assumed you wouldn't do it on purpose, since you were going to lose them.

**Comment by Cernelius** [comment]
The above is based assuming that the warning is supposed to be reliable, thus counting this issue as a bug, rather than a feature request, or anyway postponing the matter at a time in which the warning would be 100% reliable.

**Comment by Cernelius** [comment]
If it would be a vote thing, I would be with @simon33-2, in preferring (albeit not currently believing) that skipping the warning would count as proof of intentional overpurchase, if the warning was correct. Thus any time you skip a purchase warning you lose the game, unless the warning was wrong (bug).

**Comment by simon33-2** [comment]
Lose the game? Seems extreme. Being called a cheater is more what I was thinking.

Anyway the rules are for a board game where mistakes are possible.

There is one other way that an overbuy is possible without triggering the warning - if you liberate a capital (such as Paris) which causes an IC where you needed to place something to revert to its original owner. Unlikely with the US owning 3 majors and UK having so many too. But possible.

**Comment by Cernelius** [comment]
>Lose the game? Seems extreme. Being called a cheater is more what I was thinking.

Well, then you would factually end up as @panther2, if at the end you allow the refund anyways. :-)

Well, you could say that the game must be restarted right before the purchase phase when the overbuy with ignored warning happened, but not all games are PBEM and the history save is a beta feature... This would need even more to be officialised, since it would be possibly also a matter of ignoring rolled dice. And what about if at round 25 I notice an overbuy made at round 3, maybe that was not refunded (that would be forbidden nonetheless, if skipping the warning is sure proof).

**Comment by Cernelius** [comment]
>There is one other way that an overbuy is possible without triggering the warning - if you liberate a capital (such as Paris) which causes an IC where you needed to place something to revert to its original owner. Unlikely with the US owning 3 majors and UK having so many too. But possible.

Especially (or only?) if you are doing a battle to liberate that capital, would not this be a different case in which, under the dice system, you have at least a chance of not overbuying (as long as you fail to liberate), thus the purchase is legal, hence in case you cannot place all you purchase the refund is surely allowed?

**Comment by simon33-2** [comment]
True, but the refund is not supported in that case.

**Comment by panther2** [comment]
Great discussion, Guys.

@Cernelius 

> If you can confirm this, this basically means that v3 has the same issue, substantially.

You are right, this is an issue for v3, v4, v5, v6, global1940 (incl, separate theater games), as the FAQ-rule quoted above has been added to all of them.

> Are you sure about this? If I have only 1 upgradable factory and buy 2 factory upgrades, I assumed this would count as an overpurchase too. Is this correct?

The FAQ-addendum has been introduced, to prevent from selecting units to place depending on the outcome of battles before. For example you originally intended to place Armour in a territory, but as you lost a special battle you place Infantry there instead and claim a refund for the Armour, then.
This is a nice example for illegally gaining profit from an (intended) overpurchase.

Factories, Airfields and Harbours  are to be placed 
" in any eligible territory that you have
controlled since the start of your turn. Unlike combat units,
facilities do not require that an industrial complex be in the
territory in which they are mobilized, and they do not count
against the number of units that an industrial complex may
mobilize (see above) if there is one there. "
(Global Rules)

In case you buy facilities that you cannot place, you do it because you erroneously thought you could place them. You don't gain any profit from buying more than you can place, instead you are punished by losing tempo, as you did not spend that money on units instead this turn.
You can call it 'overbuy' of course, but it is another context than the case of ignoring factories' production limits on purpose as described before.

**Comment by panther2** [comment]
On a real life boardgame all of this would not be much of an issue. As the players in general would discuss it during Purchase Units phase of the "committing" player.

In case in lobby (life) games the opponents have a chance to interfer during the purchase phase, that would be quite close to the situation given in a real boardgame.

**Comment by Cernelius** [comment]
Yeah, I think there are only 3 feasible ways to solve this, all needing some developers:

1) Having the program 100% handling purchases and not allowing you to overpurchase (instead of giving just a warning). However, in this case, we would need to disable it from some games, like Napoleonic Empires, in which mobile factories allow you to overpurchase, or handling that too. It is important to keep a property to switch this off, even more so if it would be an actual limit, instead of just a warning (no reasons not allowing a custom game in which you can overpurchase, if preferred, plus there may always be new or old items, as well as triggers, that may not be covered).

2) Having the warning being 100% correct and documenting in WWII Global notes that skipping it is a violation of the rules, thus will make you lose the game, unless agreed otherwise (rewording the current notes as to tell that the refund can only happen in case you succeed to liberate a capital and you purchased for the fail). This may be better for handling custom games where you can have overpurchase due to triggers or other items, and don't want to go as far as removing the check completely (for example, Napoleonic Empires, where you have the check, but can still overpurchase if you plan to move your factories to higher production territories during NCM).

3) Having the purchase phase needing all players to validate it, before ending it; meaning that when I click on "done", the other players would get the info and he can choose to approve my purchase or unfinalize it (getting me back like I didn't click on done). I guess this would mean that, like, in a 8 players FFA we would need to wait for 7 approvals at every purchase.

Of course, the other item would be handling the refund, that I assume it would be fairly easy (?), as it just a matter of the program giving back the resource costs of the units getting destroyed, for not being placed. This doesn't matter for v3 and v4, where you just keep them, and this is already supported.

A thing I'm not sure is if in v1 and v2 you are allowed to overpurchase on purpose, and just lose what you can place, as I believe the rules were simply written under the assumption that you would never want to overpurchase. Not that much of an item, since losing what you can place is very punishing, but, in this case, those games would need remaining with the warning only check, allowing overpurchases, if that's actually allowed also when intentional (I know it doesn't actually make sense for Classic and Revised to be sanely wanting to overpurchase).

**Comment by Cernelius** [comment]
>In case in lobby (life) games the opponents have a chance to interfer during the Purchase Phase, that would be quite close to the situation given in a real boardgame.

Well... no. You get to see what the other side purchased only after the purchase phase is over, and even that is assuming you are watching the game like a hawk all the time (it would be possible to add a sound to notify a purchase has been done). Thus you could interfere only starting from the phase after the purchase one, that it is usually the combat move phase, but you can even get directly from purchase to combat, in case of move-then-buy games, that is the case of some global mods too, I believe.

**Comment by Cernelius** [comment]
Another related item is that currently the game doesn't enforce that you must place everything except what you cannot. This is actually relevant in WWIIv3 dice, where you might end up in a situation in which you would prefer to avoid placing your purchase, to place it next turn, which the game allows you to, even tho you cannot do that.

**Comment by panther2** [comment]
Good thoughts, @Cernelius 

> This doesn't matter for v3 and v4, where you just keep them, and this is already supported.

v3 only. v4 already should have the return units and get refund rule. In case TripleA handles v4 like v3, that would be wrong. (Can't test at the moment.)

> A thing I'm not sure is if in v1 and v2 you are allowed to overpurchase on purpose, and just lose what you cannot place, as I believe the rules were simply written under the assumption that you would never want to overpurchase. 

Yes, by the rules we have:

v1: units are lost, no refund
v2: units are lost, no refund
(per FAQ: Technically (which means, in a tournament), yes. In a friendly game, tell the player he's making a mistake and let him take it back or refund him the money. You don't want to win that way.)
v2-LHTR: units are lost, no refund


In v1 it may happen that you buy units that you can't place later, due to aspects happening during or because of combat. The rulebook gives some examples for it. 

v2 allows to buy as much as you can afford. When it comes to returning units you can't place the FAQ appeals to be generous to players who "make a mistake" (that sounds like the indavertent overpurchase).

I assume, too,  that the rules of v1 and v2 indeed have not been updated, because of the costly consequences when getting no refund. This should prevent from intentional overpurchase.

**Comment by simon33-2** [comment]
I hate @Cernelius 's option 3! The only change which makes sense to me is option 1.

**Comment by panther2** [comment]
While all three options are valid, I would prefer option 1, too.
This needs of course a calculation of a power's total unit production capacity, taking (possible) factory repairs into account.
(Just referencing #2219 again.)

Again, this does not solve the original issue here, where a factory upgrade was bought under wrong assumptions (https://github.com/triplea-game/triplea/issues/3506#issuecomment-402047632).

**Comment by DanVanAtta** [comment]
IIRC correctly, map XMLs specify which nations can purchase which units. If so, the simplest fix (and requested fix) here would be to remove the line that allows Japan to purchase a factory upgrade.

As @panther2 linked,  #2219 gets into the topic of how to improve this area

**Comment by DanVanAtta** [comment]
Turning something off because it does not work sounds like the definition of a work-around. What is the work-around, well, a code defect, something that is not working properly. Option 1 I believe would be basically to fix that problem. To follow up on that and track our findings in that area, I opened: https://github.com/triplea-game/triplea/issues/3769

There are some UI updates that would be also helpful here, I opened #3767 and #3768 to follow up.

Otherwise here, I think what remains would be to update the map configuration and verify that the factory upgrade is no longer available for Japan purchase

**Comment by panther2** [comment]
> IIRC correctly, map XMLs specify which nations can purchase which units. If so, the simplest fix (and requested fix) here would be to remove the line that allows Japan to purchase a factory upgrade.

> Otherwise here, I think what remains would be to update the map configuration and verify that the factory upgrade is no longer available for Japan purchase

@DanVanAtta 

In general a factory upgrade **is** available for Japan. For example the player could buy a minor factory in Korea and upgrade it, as Korea has a territory value of 3 and is originally controlled by Japan.

The original poster of this issues erroneously bought an upgrade thinking he could place it to upgrade a factroy in a territory that does not meet the requirements. He was simply missing that rule.
